src/server/search/doc_index.h

```diff
@@ -52,7 +52,102 @@ struct SearchResult {
   std::optional<facade::ErrorReply> error;
 };
 
-template <typename T> using SearchField = std::pair<T /*identifier*/, T /*short name*/>;
+enum class NameType : uint8_t { kIdentifier, kShortName, kUndefined };
+
+template <typename T> class SearchField {
+ private:
+  using SingleName = std::pair<T /*identifier or short name*/, NameType>;
+
+  static bool IsJsonPath(const T& name) {
+    if (name.size() < 2) {
+      return false;
+    }
+    return name.front() == '$' && (name[1] == '.' || name[1] == '[');
+  }
+
+ public:
+  SearchField() = default;
+
+  SearchField(T name, NameType name_type) : name_(std::make_pair(std::move(name), name_type)) {
+  }
+
+  SearchField(T name, NameType name_type, T new_alias)
+      : name_(std::make_pair(std::move(name), name_type)), new_alias_(std::move(new_alias)) {
+  }
+
+  template <typename U, typename = std::enable_if_t<std::is_constructible_v<T, U>>>
+  explicit SearchField(const SearchField<U>& other)
+      : name_(std::make_pair(T{other.name_.first}, other.name_.second)) {
+    if (other.HasNewAlias()) {
+      new_alias_ = T{other.new_alias_.value()};
+    } else {
+      new_alias_.reset();
+    }
+  }
```

Comment:
I'd suggest to simplify the code even with small sacrifices to performance, as we have now a few hundred lines dedicated to just passing around fields and aliases? We can use just plain strings everywhere, we can use dfly::StringOrView